src/cpp/shared/game/game.cpp

```diff
@@ -268,130 +267,124 @@ void game::ProcessOnMapHeroes() {
   hero *randomHero;
   char mightBeHeroAlreadyExists[56];
   int ppMapExtraHeroIdx;
-  heroMapExtra *mapExtraHero;
+  HeroExtra *mapExtraHero;
   char isJail;
   int i;
 
-  result = memset(mightBeHeroAlreadyExists, 0, 54u);// Only 54 out of 56 bytes set to zero, for whatever reason.
-  for (i = 0; i < 3; ++i) {
-    for (coordYForRandomHero = 0; ; ++coordYForRandomHero) {
-      result = (void *)MAP_HEIGHT;
-      if (coordYForRandomHero >= MAP_HEIGHT)
-        break;
-      for (coordXForRandomHero = 0; coordXForRandomHero < MAP_WIDTH; ++coordXForRandomHero) {
-        loc = &this->map.tiles[(coordYForRandomHero * this->map.width) + coordXForRandomHero];
-        if ((loc->objType & 0x7F) == 55 || loc->objType == 251) {
-          isJail = (loc->objType & 0x7F) == LOCATION_JAIL;
-          ppMapExtraHeroIdx = loc->extraInfo;
-          mapExtraHero = (heroMapExtra *)ppMapExtra[ppMapExtraHeroIdx];
-
-          if (!i) {                           // 1st iteration
-            if (!mapExtraHero->field_11 || mapExtraHero->heroID >= 54 || mightBeHeroAlreadyExists[mapExtraHero->heroID]) {
-              mapExtraHero->couldBeHasFaction = 0;
-            } else {
-              mightBeHeroAlreadyExists[mapExtraHero->heroID] = 1;
-              mapExtraHero->couldBeHasFaction = 1;
-            }
-            if (isJail) {
-              mapExtraHero->owner = -1;
-            } else {
-              mapExtraHero->owner = loc->objectIndex / 7; // Uh, redundant line?
-              mapExtraHero->owner = gcColorToPlayerPos[mapExtraHero->owner];
-            }
+  memset(mightBeHeroAlreadyExists, 0, 54u);// Only 54 out of 56 bytes set to zero, for whatever reason.
```

Comment:
56 is divisible by 4.

My guess is that mightBeHeroAlreadyExists is actually an array of length 54, but the compiler enlarged it to 56 to make these be 4-byte aligned.

Have you learned about how compilers pad things to be multiples of 4?

Reply:
I have not, but I'd imagine that it is related to the way that bytes are addressed and the concept of efficiency.

Reply:
You got it.

More importantly, in x86, you can only read a dword from an address which is divisible by 4.

Reply:
Cool.